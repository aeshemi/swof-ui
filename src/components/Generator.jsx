import React, { Component } from 'react';
import { Button, Table } from 'react-bootstrap';
import ApiClient from '../ApiClient';

class Generator extends Component {
  constructor() {
    super();
    this.state = { result: [] };
    this.generateSchedule = this.generateSchedule.bind(this);
  }

  generateSchedule() {
    ApiClient.generateSchedule().then(data => {
      this.setState({ result: data });
    });
  }

  render() {
    const { result } = this.state;

    return (
      <div>
        <Button bsStyle="primary" bsSize="large" onClick={this.generateSchedule}>
          Generate Schedule
        </Button>
        <hr />
        {result.length > 0 && (
          <Table striped>
            <thead>
              <tr>
                <th>Date</th>
                <th>Shift</th>
                <th>Engineer Id</th>
              </tr>
            </thead>
            <tbody>
              {this.state.result.map(s => {
                const dateString = s.date.substring(0, 10);

                return (
                  <tr key={`${dateString}-${s.shift}`}>
                    <td>{dateString}</td>
                    <td>{s.shift}</td>
                    <td>{s.engineerId}</td>
                  </tr>
                );
              })}
            </tbody>
          </Table>
        )}
      </div>
    );
  }
}

export default Generator;
