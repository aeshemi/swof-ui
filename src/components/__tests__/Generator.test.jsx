import Generator from '../Generator';

describe('Generator', () => {
  it('renders correctly on init', () => {
    const wrapper = shallow(<Generator />);
    expect(wrapper).toMatchSnapshot();
  });

  it('click on generate calls generateSchedule and returns results', done => {
    fetch.mockResponse(
      JSON.stringify([
        { date: '2018-08-20T00:00:00', shift: 'Morning', engineerId: 6 },
        { date: '2018-08-20T00:00:00', shift: 'Afternoon', engineerId: 7 },
        { date: '2018-08-21T00:00:00', shift: 'Morning', engineerId: 10 },
        { date: '2018-08-21T00:00:00', shift: 'Afternoon', engineerId: 5 },
        { date: '2018-08-22T00:00:00', shift: 'Morning', engineerId: 4 },
        { date: '2018-08-22T00:00:00', shift: 'Afternoon', engineerId: 9 },
        { date: '2018-08-23T00:00:00', shift: 'Morning', engineerId: 8 },
        { date: '2018-08-23T00:00:00', shift: 'Afternoon', engineerId: 3 },
        { date: '2018-08-24T00:00:00', shift: 'Morning', engineerId: 1 },
        { date: '2018-08-24T00:00:00', shift: 'Afternoon', engineerId: 2 }
      ])
    );

    const spy = jest.spyOn(Generator.prototype, 'generateSchedule');
    const wrapper = shallow(<Generator />);

    wrapper.find('Button').simulate('click');

    setImmediate(() => {
      wrapper.update();

      expect(spy).toHaveBeenCalled();
      expect(wrapper).toMatchSnapshot();

      spy.mockReset();
      spy.mockRestore();

      fetch.resetMocks();

      done();
    });
  });
});
