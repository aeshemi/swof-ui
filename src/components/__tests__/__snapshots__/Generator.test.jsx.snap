// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generator click on generate calls generateSchedule and returns results 1`] = `
<div>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsSize="large"
    bsStyle="primary"
    disabled={false}
    onClick={[Function]}
  >
    Generate Schedule
  </Button>
  <hr />
  <Table
    bordered={false}
    bsClass="table"
    condensed={false}
    hover={false}
    responsive={false}
    striped={true}
  >
    <thead>
      <tr>
        <th>
          Date
        </th>
        <th>
          Shift
        </th>
        <th>
          Engineer Id
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        key="2018-08-20-Morning"
      >
        <td>
          2018-08-20
        </td>
        <td>
          Morning
        </td>
        <td>
          6
        </td>
      </tr>
      <tr
        key="2018-08-20-Afternoon"
      >
        <td>
          2018-08-20
        </td>
        <td>
          Afternoon
        </td>
        <td>
          7
        </td>
      </tr>
      <tr
        key="2018-08-21-Morning"
      >
        <td>
          2018-08-21
        </td>
        <td>
          Morning
        </td>
        <td>
          10
        </td>
      </tr>
      <tr
        key="2018-08-21-Afternoon"
      >
        <td>
          2018-08-21
        </td>
        <td>
          Afternoon
        </td>
        <td>
          5
        </td>
      </tr>
      <tr
        key="2018-08-22-Morning"
      >
        <td>
          2018-08-22
        </td>
        <td>
          Morning
        </td>
        <td>
          4
        </td>
      </tr>
      <tr
        key="2018-08-22-Afternoon"
      >
        <td>
          2018-08-22
        </td>
        <td>
          Afternoon
        </td>
        <td>
          9
        </td>
      </tr>
      <tr
        key="2018-08-23-Morning"
      >
        <td>
          2018-08-23
        </td>
        <td>
          Morning
        </td>
        <td>
          8
        </td>
      </tr>
      <tr
        key="2018-08-23-Afternoon"
      >
        <td>
          2018-08-23
        </td>
        <td>
          Afternoon
        </td>
        <td>
          3
        </td>
      </tr>
      <tr
        key="2018-08-24-Morning"
      >
        <td>
          2018-08-24
        </td>
        <td>
          Morning
        </td>
        <td>
          1
        </td>
      </tr>
      <tr
        key="2018-08-24-Afternoon"
      >
        <td>
          2018-08-24
        </td>
        <td>
          Afternoon
        </td>
        <td>
          2
        </td>
      </tr>
    </tbody>
  </Table>
</div>
`;

exports[`Generator renders correctly on init 1`] = `
<div>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsSize="large"
    bsStyle="primary"
    disabled={false}
    onClick={[Function]}
  >
    Generate Schedule
  </Button>
  <hr />
</div>
`;
